### PR TITLE
Refactor/external communication

### PR DESCRIPTION
**Pull request by devceline**
> *Prefer composition over inheritance*

# Changes
- Move from an inheritance based approach for external communicators to composition based *Communicator* modules. 
- This simplifies the code from having duplicate android/ios communicators
- Nothing changes with how communication actually happens, just moved code around

### CODE REVIEW

**Comment by bkrem** [comment]
Could you update the PR description with any relevant context pls? 

Overall seems to be consolidating the provider pattern usage to be more DRY as we talked about a while back which is great, is there anything else to look out for during review?

**Comment by devceline** [comment]
> Could you update the PR description with any relevant context pls?
> 
> Overall seems to be consolidating the provider pattern usage to be more DRY as we talked about a while back which is great, is there anything else to look out for during review?

Updated PR description

File: src/w3iProxy/pushProviders/internalPushProvider.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ export default class InternalPushProvider implements W3iPushProvider {
   private readonly emitter: EventEmitter
   public providerName = 'InternalChatProvider'
 
-  public constructor(emitter: EventEmitter) {
+  public constructor(emitter: EventEmitter, _name = 'internal') {
```
Comment (bkrem) [inline src/w3iProxy/pushProviders/internalPushProvider.ts:11]: Guess I'm missing something but I'm actually not clear on the `_` prefix here, why not just do default assignment as `name = 'internal'`?

**Review by Cali93 — APPROVED** [review]
LGTM, feels cleaner this way 👌

File: src/w3iProxy/w3iPushFacade.ts [hunk 2]
```diff
@@ -9,17 +9,15 @@ import type {
...
-import iOSPushProvider from './pushProviders/iosPushProvider'
 import InternalPushProvider from './pushProviders/internalPushProvider'
 import type { PushFacadeEvents } from './listenerTypes'
 
 class W3iPushFacade implements W3iPush {
   private readonly providerMap = {
     internal: InternalPushProvider,
     external: ExternalPushProvider,
-    android: AndroidPushProvider,
-    ios: iOSPushProvider
```
Comment (bkrem) [inline src/w3iProxy/w3iPushFacade.ts:22]: Can we also delete the `iOSPushProvider`?

File: src/w3iProxy/pushProviders/externalPushProvider.ts [hunk 3]
```diff
@@ -2,18 +2,34 @@ import type { EventEmitter } from 'events'
...
+      case 'ios':
+        this.communicator = new IOSCommunicator(this.emitter)
+        break
+      default:
+        this.communicator = new JsCommunicator(this.emitter)
+        break
+    }
   }
 
   protected async postToExternalProvider<MName extends keyof PushClientFunctions>(
```
Comment (bkrem) [inline src/w3iProxy/pushProviders/externalPushProvider.ts:35]: Should this also delegate to the `communicator` here as for `externalChatProvider`? https://github.com/WalletConnect/web3inbox/pull/51/files#diff-1f7b1977dc91a5064fee19ccb242c192b43705ceb88616ca0c87a7ad61a8091eR48

**Review by bkrem — APPROVED** [review]
✨

File: src/w3iProxy/pushProviders/internalPushProvider.ts (hunk 1 above)
Comment (devceline) [reply src/w3iProxy/pushProviders/internalPushProvider.ts:11]: Simply to denote it's not used anywhere

File: src/w3iProxy/w3iPushFacade.ts (hunk 2 above)
Comment (devceline) [reply src/w3iProxy/w3iPushFacade.ts:22]: There is no `iOSPushProvider`

**Pull request merged**